src/main/java/algorithms/binarySearch/binarySearchTemplated/README.md

```diff
@@ -0,0 +1,119 @@
+# BinarySearchTemplated
+
+## Background
+
+BinarySearchTemplated is a more generalised algorithm of [BinarySearch](../binarySearch) that removes the condition that
+checks if the current mid-value is equal to the target (which helps to end the search the moment the target is found). 
+The template adds a "condition" method which will be modified based on the requirements of the implementation.
+
+The narrowing of the search space differs from BinarySearch - only the high pointer will be adjusted by one unit.
+
+This template will work for most binary search problems and will require the following changes when used for different 
+problems:
+- Search space (high and low pointers)
+- Condition method
+- Returned value (low or low - 1)
+
+### Search Space (Requires change)
+Initialise the boundary values of the high and low pointers to include all possible elements in the search space.
+
+### Condition (Requires change)
+We assume that when the condition returns true, the current value "passes" and when the condition returns false, the
+current value "fails".
+
+Note that in this template, the conditional blocks
+```
+if (condition(x)) {
+   high = mid;
+} else {
+   low = mid + 1;
+}
+```
+requires elements that "fail" the condition to be on the left of the elements that "pass" the condition, see below, in a
+sorted array due to the way the high and low pointers are reassigned.
+
+![binary search templated 1 img](../../../../../../docs/assets/images/BinarySearchTemplated1.jpeg)
+
+Hence, we will need to implement a condition method that is able to discern between arrays that "pass" and "fail"
+accurately and also place them in the correct relative positions i.e. "fail" on the left of "pass". Suppose we change
+the condition method implementation in BinarySearchTemplated from `value >= target` to `value < target`, what will
```

Comment:
possible to dissociate from implementation-specific like value >= target? Since it is not always the case that condition function involves comparison of numeric values.

Maybe can phrase it smth like: Occasionally, it might make more sense to construct the condition function such that we have the following structure: P P P F F F F F. Finding the first "F" is similar and requires only these changes: ....

Reply:
or
what if our condition is structured like: P P P F F F.., what will happen?